llama_index/vector_stores/qdrant_utils.py

```diff
@@ -0,0 +1,132 @@
+from typing import Any, Callable, List, Protocol, Tuple, runtime_checkable
+
+from llama_index.vector_stores.types import VectorStoreQueryResult
+
+SparseEncoderCallable = Callable[[List[str]], Tuple[List[List[int]], List[List[float]]]]
+
+
+@runtime_checkable
+class HybridFusionCallable(Protocol):
+    """Hybrid fusion callable protocol."""
+
+    def __call__(
+        self,
+        dense_result: VectorStoreQueryResult,
+        sparse_result: VectorStoreQueryResult,
+        **kwargs: Any,
+    ) -> VectorStoreQueryResult:
+        """Hybrid fusion callable."""
+        ...
+
+
+def default_sparse_encoder(model_id: str) -> SparseEncoderCallable:
+    try:
+        import torch
+        from transformers import AutoModelForMaskedLM, AutoTokenizer
+    except ImportError:
+        raise ImportError(
+            "Could not import transformers library. "
+            'Please install transformers with `pip install "transformers[torch]"`'
+        )
+
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+    model = AutoModelForMaskedLM.from_pretrained(model_id)
+    if torch.cuda.is_available():
+        model = model.to("cuda")
+
+    def compute_vectors(texts: List[str]) -> Tuple[List[List[int]], List[List[float]]]:
+        """
+        Computes vectors from logits and attention mask using ReLU, log, and max operations.
+        """
+        # TODO: compute sparse vectors in batches if max length is exceeded
+        tokens = tokenizer(texts, truncation=True, padding=True, return_tensors="pt")
+        if torch.cuda.is_available():
+            tokens = tokens.to("cuda")
+
+        output = model(**tokens)
+        logits, attention_mask = output.logits, tokens.attention_mask
+        relu_log = torch.log(1 + torch.relu(logits))
+        weighted_log = relu_log * attention_mask.unsqueeze(-1)
+        tvecs, _ = torch.max(weighted_log, dim=1)
+
+        # extract the vectors that are non-zero and their indices
+        indices = []
+        vecs = []
+        for batch in tvecs:
+            indices.append(batch.nonzero(as_tuple=True)[0].tolist())
+            vecs.append(batch[indices[-1]].tolist())
+
+        return indices, vecs
+
+    return compute_vectors
+
+
+def relative_score_fusion(
+    dense_result: VectorStoreQueryResult,
+    sparse_result: VectorStoreQueryResult,
+    alpha: float = 0.5,
+    top_k: int = 2,
+) -> VectorStoreQueryResult:
+    """
+    Fuse dense and sparse results using relative score fusion.
+    """
+    # sanity check
+    assert dense_result.nodes is not None
+    assert dense_result.similarities is not None
+    assert sparse_result.nodes is not None
+    assert sparse_result.similarities is not None
```

Comment:
in some corner cases the results of one side is empty, we could stop here
could we make  a check before dedup or fusion. if any one list is empty, we just return the other list

Reply:
I thiiiink that will not be the case? At least with qdrant, these will always be populated.

Good to note though, in case we expand this approach to other vector dbs